openedx/core/djangoapps/course_roles/helpers.py

```diff
@@ -0,0 +1,172 @@
+    if not use_permission_checks():
+        return False
+    elif isinstance(user, AnonymousUser):
+        return False
+    if isinstance(permission, CourseRolesPermission):
+        permission = permission.value.name
+    return UserRole.objects.filter(
+        user=user,
+        role__permissions__name=permission,
+        course__isnull=True,
+        org__name=organization_name,
+    ).exists()
+
+
+def user_has_permission_list_org(
+        user: Union[User, AnonymousUser],
+        permissions: List[Union[CourseRolesPermission, str]],
+        organization_name: str
+):
+    """
+    Check if a user has ALL of the given permissions for all courses in an organization.
+    """
+    if not use_permission_checks():
+        return False
+    return all(
+        user_has_permission_org(user, permission, organization_name) for permission in permissions
+    )
+
+
+def user_has_permission_course_org(
+        user: Union[User, AnonymousUser],
+        permission: Union[CourseRolesPermission, str],
+        course_key: CourseKey
+):
+    """
+    Check if a user has a permission for all courses in an organization or for a specific course.
+    """
+    if not use_permission_checks():
+        return False
+    elif isinstance(user, AnonymousUser):
+        return False
+    if isinstance(permission, CourseRolesPermission):
+        permission = permission.value.name
+    return (user_has_permission_course(user, permission, course_key) or
+            user_has_permission_org(user, permission, course_key.org)
+            )
+
+
+def user_has_permission_list_course_org(
+        user: Union[User, AnonymousUser],
+        permissions: List[Union[CourseRolesPermission, str]],
+        course_key: CourseKey
+):
+    """
+    Check if a user has all of the given permissions for all courses in an organization or for a specific course.
+    """
+    if not use_permission_checks():
+        return False
+    return all(
+        user_has_permission_course_org(user, permission, course_key)
+        for permission in permissions
+    )
+
+
+def get_all_user_permissions_for_a_course(user_id: int, course_key: CourseKey):
+    """
+    Get all of a user's permissions for a course,
+    including, if applicable, organization-wide permissions
+    and instance-wide permissions.
+    """
+    if user_id is None:
+        raise ValueError('user_id must not be None')
+    if course_key is None:
+        raise ValueError('course_key must not be None')
+    try:
+        user = User.objects.get(pk=user_id)
+    except User.DoesNotExist as exc:
+        raise ValueError('user does not exist') from exc
+    if not CourseOverview.course_exists(course_key):
+        raise ValueError('course does not exist')
+    permissions = set(UserRole.objects.filter(
+        Q(Q(user__id=user_id) &
+          Q(Q(course=course_key) | Q(course__isnull=True) & Q(Q(org__name=course_key.org) | Q(org__isnull=True))))
+    ).values_list('role__permissions__name', flat=True))
```

Comment:
Arguments to `filter` are implicitly AND'd together. You can also chain together queries without wrapping `Q`s. So I'm not sure if I got the parens all right, but maybe something like this?

```python
permissions_qset = UserRole.objects.filter(
  Q(user__id=user_id),
  (
      # Course-specific roles
      Q(course=course_key) |

      # Org-wide roles that apply to this course
      (Q(course__isnull=True) & Q(org__name=course.org)) |

      # System-wide roles
      Q(org__isnull=True)
  )
)
permissions = set(
    permissions_qset.values_list('role__permissions__name', flat=True)
)
```